src/storage/src/source/persist_source.rs

```diff
@@ -59,172 +56,34 @@ pub fn persist_source<G>(
-                            // in one go.
-                            //
-                            // NOTE: We use this seemingly complicated approach
-                            // such that the code is ready to deal with
-                            // partially ordered times.
-                            for cap in cap_set.iter() {
-                                // NOTE: The nightly `drain_filter()` would use
-                                // less allocations than this. Should switch to
-                                // it once it's available in stable rust.
-                                let (mut to_emit, remaining_updates) =
-                                    updates.into_iter().partition(|(_update, ts, _diff)| {
-                                        PartialOrder::less_equal(cap.time(), ts)
-                                    });
-                                updates = remaining_updates;
-
-                                let mut session = output.session(&cap);
-                                session.give_vec(&mut to_emit);
-                            }
-
-                            assert!(
-                                updates.is_empty(),
-                                "did not have matching Capability for updates: {:?}",
-                                updates
-                            );
-                        }
-                        Some(Err::<_, ExternalError>(e)) => {
-                            // TODO(petrosagg): error handling
-                            panic!("unexpected error from persist {e}");
-                        }
-                        None => {
-                            // Empty out the `CapabilitySet` to indicate that we're done.
-                            cap_set.downgrade(&[]);
-                            return;
-                        }
-                    }
-                }
-            }
-        });
-
-    let (ok_stream, err_stream) = timely_stream.ok_err(|x| match x {
-        ((Ok(SourceData(Ok(row))), Ok(())), ts, diff) => Ok((row, ts, diff)),
-        ((Ok(SourceData(Err(err))), Ok(())), ts, diff) => Err((err, ts, diff)),
-        // TODO(petrosagg): error handling
-        _ => panic!("decoding failed"),
-    });
-
-    let token = Rc::new(token);
-
-    (ok_stream, err_stream, token)
-}
+    // This source is split as such:
+    // 1. Sets up `async_stream`, which only yields data (hollow batches) on one
+    //    chosen worker. Generating hollow batches also generates `SeqNo` leases
+    //    on the chosen workr.
+    // 2. Batch distribution: A timely source operator which continuously reads
+    //    from that stream, and distributes the data among workers.
+    // 3. Batch fetcher: A timely operator which downloads the batch's contents
+    //    from S3, and outputs them to a timely stream. Additionally, the
+    //    operator returns the `SeqNo` lease to the original worker.
```

Comment:
Needs to be changed based on discussion around the lease/acknowledge API.

Reply:
Done in latest push